arrow-core-data/src/main/kotlin/arrow/core/Option.kt

```diff
@@ -495,47 +712,254 @@ sealed class Option<out A> : OptionOf<A> {
    *
    * @param predicate the predicate used for testing.
    */
-  fun filterNot(predicate: Predicate<A>): Option<A> =
+  inline fun filterNot(predicate: (A) -> Boolean): Option<A> =
     flatMap { a -> if (!predicate(a)) Some(a) else None }
 
   /**
    * Returns true if this option is nonempty '''and''' the predicate
    * $p returns true when applied to this $option's value.
    * Otherwise, returns false.
    *
+   * Example:
+   * ```
+   * Some(12).exists { it > 10 } // Result: true
+   * Some(7).exists { it > 10 }  // Result: false
+   *
+   * val none: Option<Int> = None
+   * none.exists { it > 10 }      // Result: false
+   * ```
+   *
    * @param predicate the predicate to test
    */
-  fun exists(predicate: Predicate<A>): Boolean = fold({ false }, { a -> predicate(a) })
+  inline fun exists(predicate: (A) -> Boolean): Boolean = fold({ false }, predicate)
+
+  inline fun filterA(predicate: (A) -> Iterable<Boolean>): Iterable<Option<A>> =
+    traverseFilter { a -> predicate(a).map { if (it) Some(a) else None } }
+
+  inline fun <AA> filterAEither(predicate: (A) -> Either<AA, Boolean>): Either<AA, Option<A>> =
+    traverseFilterEither { a -> predicate(a).map { if (it) Some(a) else None } }
+
+  inline fun <AA> filterAValidated(predicate: (A) -> Validated<AA, Boolean>): Validated<AA, Option<A>> =
+    traverseFilterValidated { a -> predicate(a).map { if (it) Some(a) else None } }
 
   /**
    * Returns true if this option is empty '''or''' the predicate
    * $p returns true when applied to this $option's value.
    *
    * @param p the predicate to test
    */
-  fun forall(p: Predicate<A>): Boolean = fold({ true }, p)
+  inline fun forall(p: (A) -> Boolean): Boolean = fold({ true }, p)
```

Comment:
This is named `all` in Kotlin, https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/all.html.
We should deprecate towards `all`.